Automattic/vip-go-mu-plugins
# Cron: Action Scheduler Dynamic Queue

## Description

### What is Action Scheduler

[Action Scheduler](https://actionscheduler.org/) is an event-runner that can be heavily utilized by plugins like WooCommerce. "Actions" are scheduled and stored in the database, and then the "queue runner" picks up the actions in batches and processes through them. It is similar to how our cron works, except it more efficiently handles multiple actions within one request - making it ideal for running hundreds to thousands of "small" scheduled tasks (like order webhooks, emails, inventory syncs, etc).

The way the queue runner works by default in AS is through a recurring 60second cron job called `action_scheduler_run_queue`. It will claim and process say a batch of 25 actions repeatedly as long as it can given its time limit of 30 seconds. After which it will wait until the next iteration of the recurring cron job to go again. This alone has the main shortcoming of being pretty limited on concurrency.

There is also an "async runner" in the core plugin, which kicks off a queue once a minute off of a frontend request (`shutdown`). This is meant as a solution to normal WP cron being lacking, but it's not really necessary on our infrastructure - would be much better if we just let cron containers handle the background tasks.

Even with the async runner, it's still very easy for the queue to become backed up w/ high volume sites. Subscriptions are a common use case where thousands of tasks can become due all at once.

### Proposed Solution

Our cron runner is very well equipped to kick off multiple jobs at once, so let's utilize it :). This PR disables the async frontend queue in favor of scheduling one-time jobs when they are needed. If a high volume of actions hit or if the default recurring job starts to fall behind, this dynamic queue will jump in and dispatch additional cron jobs that are allowed to run concurrently with each other.

Rather than needing to tailor a bunch of filters per a site's needs, this scales directly off of `JOB_CONCURRENCY_LIMIT`. Allowing up to 33% of the concurrency limit. If that fails to be sufficient, then the site should have the concurrency limit increased at the application level at that point. But also notably, it is very lightweight in times where all the extra queues are not needed. So rather than having an unnecessary amount of recurring jobs that are always filling up the cron queue, it will only schedule jobs when it both needs them and when action scheduler has slots available and is ready to launch new queues.

Additionally, the default 30-second queue timeout limit is there because general WP hosts run cron on the frontend and thus it is limited by timeouts there. We can push this much farther on our cron containers, however we do not want to go too far as memory leaks become a concern if a single job tries process too many actions. So settling at a fairly-well tested 120seconds for now. This gives the default recurring queue and the additional queues a lot of efficiency, and lowers how often we need to worry about dispatching new queues.

## Changelog Description

### Cron: Introduce Dynamic Queue system for Action Scheduler

This new feature tailors Action Scheduler defaults and unleashes more processing power by syncing up with the Cron Control system. It will schedule additional AS queues to run concurrently when they are needed to process through a large list of "due" actions.

## Checklist

Please make sure the items below have been covered before requesting a review:


## Steps to Test

- Install this file as a client-mu-plugin and deploy to a test site, as you'll need it running live on real batch containers.
- Also add in this mini plugin, which will make it easy to set up a large backlog of actions: https://gist.github.com/WPprodigy/aebd49bc2f7f10bcb62dbe342948f924

1) Schedule 2000 actions that will run for 1 second each: `wp vip-action-scheduler-jammer --amount=2000 --duration=1`
2) Watch as over time the "In-progress" list of actions starts to increase, showing the concurrency: `/wp-admin/admin.php?page=wc-status&tab=action-scheduler&status=in-progress`. Can also `wp cron-control events list` and watch for the new events to be added (note that the dispatch is cached, so may need to wait for around a minute - else comment out the cache check & load a page on the sandbox).

(test site 3474 is available and setup for this 🙂 )
